cmd/state-handler/main.go

```diff
@@ -0,0 +1,126 @@
+	kubeconfig    = flag.String("kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	master        = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	namespace     = flag.String("n", "", "The namespace where the CRDs are created. If left blank and running via pod, it will be taken from there.")
+	hostname      = flag.String("host", "", "Name of the host on which to enforce and report state. If left blank and running via pod, it will be taken from there.")
+)
+
+func main() {
+	flag.Parse()
+
+	switch *executionType {
+	case "":
+		panic("execution-type must be specified")
+	case "controller":
+		controller()
+	case "client":
+		client()
+	}
+}
+
+func controller() {
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(*master, *kubeconfig)
+	if err != nil {
+		klog.Fatalf("Error building kubeconfig: %s", err.Error())
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		klog.Fatalf("Error building kubernetes clientset: %s", err.Error())
+	}
+
+	nmstateClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		klog.Fatalf("Error building nmstate clientset: %s", err.Error())
+	}
+
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, time.Second*30)
+	nmstateInformerFactory := informers.NewSharedInformerFactory(nmstateClient, time.Second*30)
+
+	ns := utils.GetNamespace(*namespace)
+	controller := NewController(
+		kubeClient,
+		nmstateClient,
+		nmstateInformerFactory.Nmstate().V1().NodeNetworkStates(),
+		utils.GetHostName(*hostname, cfg, ns),
+		ns,
+	)
+
+	// notice that there is no need to run Start methods in a separate goroutine. (i.e. go kubeInformerFactory.Start(stopCh)
+	// Start method is non-blocking and runs all registered informers in a dedicated goroutine.
+	kubeInformerFactory.Start(stopCh)
+	nmstateInformerFactory.Start(stopCh)
+
+	if err = controller.Run(2, stopCh); err != nil {
+		klog.Fatalf("Error running controller: %s", err.Error())
+	}
+}
+
+func client() {
+	flag.Parse()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(*master, *kubeconfig)
+	if err != nil {
+		klog.Fatalf("Error building kubeconfig: %v\n", err)
+	}
+
+	nmstateClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		klog.Fatalf("Error building nmstate clientset: %v\n", err)
+	}
+
+	// get name space even if not set as commandline parameter
+	ns := utils.GetNamespace(*namespace)
+
+	list, err := nmstateClient.NmstateV1().NodeNetworkStates(ns).List(metav1.ListOptions{})
+	if err != nil {
+		klog.Fatalf("Error listing all node network states (in %s): %v\n", ns, err)
+	}
+
+	nodeFound := false
+	name := utils.GetHostName(*hostname, cfg, ns)
+	if name == "" {
+		klog.Fatalf("Failed to get host name\n")
+	}
+
+	for _, state := range list.Items {
```

Comment:
inside the controller code, don't we have similar initialization to the one we perform here?
or is it all based on getting updated from k8s?

Reply:
oh, missed this one. will fix it in a second

Reply:
I don't see how could we merge the logic at this point. I need to research more about the controller. Could we take this PR as it is?